utcp.c

```diff
@@ -1131,33 +1442,31 @@ struct timeval utcp_timeout(struct utcp *utcp) {
 			continue;
 		}
 
-		if(timerisset(&c->conn_timeout) && timercmp(&c->conn_timeout, &now, <)) {
-			errno = ETIMEDOUT;
-			c->state = CLOSED;
-			if(c->recv)
-				c->recv(c, NULL, 0);
-			continue;
+		// check connection timeout
+		if(timerisset(&c->conn_timeout)) {
+			 if(timercmp(&c->conn_timeout, &now, <)) {
+				errno = ETIMEDOUT;
+				c->state = CLOSED;
+				if(c->recv)
+					c->recv(c, NULL, 0);
+				continue;
+			}
+			if(timercmp(&c->conn_timeout, &next, <))
```

Comment:
On Thu, Oct 29, 2015 at 07:41:12AM -0700, vollkorn1982 wrote:

> > +			if(timercmp(&c->conn_timeout, &next, <))
> 
> Hm. `timercmp` is posix only as well, isn't it?

Good point. Actually, according to the manpage, it's not even POSIX, but
rather a BSD thing.

I should rewrite everything to use clock_gettime(), since that is the
POSIXest way to do time nowadays. Also, clock_gettime(CLOCK_MONOTONIC)
is safer to use, and perhaps even CLOCK_MONOTONIC_COARSE can be used in
UTCP.

There is no standard C99 function to get time with a resolution better
than a second, so for Windows we need to write our own clock_gettime(),
like suggested here:

http://stackoverflow.com/questions/5404277/porting-clock-gettime-to-windows

-- 
Met vriendelijke groet / with kind regards,
     Guus Sliepen <guus@tinc-vpn.org>